### PR TITLE
[IMP] mail, hr: constraints when channel_type is not 'channel'

### PR DESCRIPTION
**Pull request by zel-odoo**
The group_public_id and group_id and subscription_department_ids
should only be used when the channel_type is 'channel'.

task-2766712

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/models/mail_channel.py [hunk 1]
```diff
@@ -89,21 +89,31 @@ def _generate_random_token(self):
...
+    def _constraint_group_id_chat(self):
+        for ch in self.sudo().filtered(lambda ch: ch.channel_type != 'channel'):
+            if ch.group_public_id or ch.group_ids:
+                raise ValidationError(_("A channel of type 'chat'/'group' should not have constraints on the public group settings."))
+
     # COMPUTE / INVERSE
+    @api.depends('channel_type')
+    def _compute_group_public_id(self):
+        for record in self:
+            record.group_public_id = self.env.ref('base.group_user') if record.channel_type == 'channel' else None
```
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/models/mail_channel.py:116]: Also I think the `ref()` in loop might be costly, like making many queries... It would be better to save it into a variable before the loop.

File: addons/mail/models/mail_channel.py [hunk 2]
```diff
@@ -89,21 +89,31 @@ def _generate_random_token(self):
...
-    # CHAT CONSTRAINT
+    # CONSTRAINTS
 
     @api.constrains('channel_last_seen_partner_ids', 'channel_partner_ids')
     def _constraint_partners_chat(self):
         for ch in self.sudo().filtered(lambda ch: ch.channel_type == 'chat'):
             if len(ch.channel_last_seen_partner_ids) > 2 or len(ch.channel_partner_ids) > 2:
                 raise ValidationError(_("A channel of type 'chat' cannot have more than two users."))
 
+    @api.constrains('group_public_id', 'group_ids')
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:106]: Note that technically it also depends on `channel_type` (and the constraint above too). But as it should never change in practice and we want to actually prevent from being able to change it in a future PR, I guess it's acceptable to omit it here.

File: addons/mail/models/mail_channel.py [hunk 3]
```diff
@@ -89,21 +89,31 @@ def _generate_random_token(self):
...
+    def _constraint_group_id_chat(self):
+        for ch in self.sudo().filtered(lambda ch: ch.channel_type != 'channel'):
+            if ch.group_public_id or ch.group_ids:
+                raise ValidationError(_("A channel of type 'chat'/'group' should not have constraints on the public group settings."))
+
     # COMPUTE / INVERSE
+    @api.depends('channel_type')
+    def _compute_group_public_id(self):
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:114]: guideline: have the compute method sorted among them the same way as the corresponding fields

(also keep the empty line below the # title)

**Review by seb-odoo — CHANGES_REQUESTED** [review]
For the commit, make sure to put a space between `[IMP]mail`, and also to keep lines below 80 chars.

Also double check enterprise rebase, the branches are currently unsync according to runbot.

And some clean up comments below.

And finally, you also need to add a similar constraint in `hr` for `subscription_department_ids` (this is similar to user group but for employees).

PS: I know you are already working on it, but we'd like an upgrade script to clean pre-existing inconsistencies.

File: addons/hr/models/mail_channel.py [hunk 4]
```diff
@@ -11,6 +12,14 @@ class Channel(models.Model):
         'hr.department', string='HR Departments',
         help='Automatically subscribe members of those departments to the channel.')
 
+    # CONSTRAINTS
+
+    @api.constrains('subscription_department_ids')
+    def _constraint_subscription_department_ids_channel(self):
+        for channel in self.sudo().filtered(lambda channel: channel.channel_type != 'channel'):
+            if channel.subscription_department_ids:
+                raise ValidationError(_('Department auto-subscription is only supported on channels. Current channel type: %s', channel.channel_type))
```
Comment (seb-odoo) [inline addons/hr/models/mail_channel.py:21]: Ideally use named parameters for translations.

File: addons/mail/models/mail_channel.py [hunk 5]
```diff
@@ -88,21 +88,26 @@ def _generate_random_token(self):
...
     def _constraint_partners_chat(self):
         for ch in self.sudo().filtered(lambda ch: ch.channel_type == 'chat'):
             if len(ch.channel_last_seen_partner_ids) > 2 or len(ch.channel_partner_ids) > 2:
                 raise ValidationError(_("A channel of type 'chat' cannot have more than two users."))
 
+    @api.constrains('group_public_id', 'group_ids')
+    def _constraint_group_id_channel(self):
+        for channel in self.sudo().filtered(lambda channel: channel.channel_type != 'channel'):
+            if channel.group_public_id or channel.group_ids:
+                raise ValidationError(_('Group authorization and group auto-subscription are only supported on channels. Current channel type: %s', channel.channel_type))
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:109]: named translation

File: addons/mail/models/mail_channel.py [hunk 6]
```diff
@@ -88,21 +88,26 @@ def _generate_random_token(self):
...
     @api.constrains('channel_last_seen_partner_ids', 'channel_partner_ids')
     def _constraint_partners_chat(self):
         for ch in self.sudo().filtered(lambda ch: ch.channel_type == 'chat'):
             if len(ch.channel_last_seen_partner_ids) > 2 or len(ch.channel_partner_ids) > 2:
                 raise ValidationError(_("A channel of type 'chat' cannot have more than two users."))
 
+    @api.constrains('group_public_id', 'group_ids')
+    def _constraint_group_id_channel(self):
+        for channel in self.sudo().filtered(lambda channel: channel.channel_type != 'channel'):
+            if channel.group_public_id or channel.group_ids:
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:108]: The `group_public_id` constraint could be made in SQL too for a stronger guarantee.

Not the other one unfortunately as it is in another table, but it's less important. If we really want, we can add a check in `_subscribe_users_automatically` to ensure the `channel_type` is as expected, but that's probably overkill?

File: addons/mail/models/mail_channel.py [hunk 7]
```diff
@@ -88,21 +88,26 @@ def _generate_random_token(self):
         ('groups', 'Selected group of users')], string='Privacy',
         required=True, default='groups',
         help='This group is visible by non members. Invisible groups can add members through the invite button.')
-    group_public_id = fields.Many2one('res.groups', string='Authorized Group',
-                                      default=lambda self: self.env.ref('base.group_user'))
+    group_public_id = fields.Many2one('res.groups', string='Authorized Group', compute='_compute_group_public_id', readonly=False, store=True)
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:91]: Ideally this should actually be readonly=True when channel type is invalid to prevent the user from inputting invalid values that will be refused by the constraint. Could possibly add the domain here in python, or at least in the form view.

File: addons/mail/models/mail_channel.py (hunk 6 above)
Comment (zel-odoo) [reply addons/mail/models/mail_channel.py:108]: yes, we can do that but maybe it is not mandatory? cause this has covered most cases and the more we have the worse performance we will get?

for the group_public_id, I have the same doubt as well. do we really need the SQL constraint?

File: addons/mail/__manifest__.py [hunk 8]
```diff
@@ -2,7 +2,7 @@
 
 {
     'name': 'Discuss',
-    'version': '1.7',
+    'version': '1.8',
```
Comment (tde-banana-odoo) [inline addons/mail/__manifest__.py:5]: Don't increase when you have no model modification. Version increase in SM apps is done only when we really modify models; here it is only data cleaning in DB.

File: addons/mail/models/mail_channel.py [hunk 9]
```diff
@@ -88,21 +88,26 @@ def _generate_random_token(self):
...
 
     @api.constrains('channel_last_seen_partner_ids', 'channel_partner_ids')
     def _constraint_partners_chat(self):
         for ch in self.sudo().filtered(lambda ch: ch.channel_type == 'chat'):
             if len(ch.channel_last_seen_partner_ids) > 2 or len(ch.channel_partner_ids) > 2:
                 raise ValidationError(_("A channel of type 'chat' cannot have more than two users."))
 
+    @api.constrains('group_public_id', 'group_ids')
+    def _constraint_group_id_channel(self):
+        for channel in self.sudo().filtered(lambda channel: channel.channel_type != 'channel'):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_channel.py:107]: Either do an "any" to raise at first found channel, either aggregate failing channels and add their name in the error (both are done in Odoo).

File: addons/mail/models/mail_channel.py (hunk 5 above)
Comment (tde-banana-odoo) [inline addons/mail/models/mail_channel.py:109]: Your current error does not add valuable information as we don't know which channel is failing. Knowing the current channel type is not really interesting to fix the issue. I would say something like ``_("The following channels have group-based authorization but are not of right type 'channel': %(channel_names)s"), channel_names = ', '.join(failing_channel.mapped('name'))``

File: addons/mail/models/mail_channel.py [hunk 10]
```diff
@@ -168,6 +173,12 @@ def _compute_member_count(self):
         for channel in self:
             channel.member_count = member_count_by_channel_id.get(channel.id, 0)
 
+    @api.depends('channel_type')
+    def _compute_group_public_id(self):
+        group_user = self.env.ref('base.group_user')
+        for channel in self:
+            channel.group_public_id = group_user if channel.channel_type == 'channel' else None
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_channel.py:180]: You should not erase a previously set value. I would do something less invasive and batch-enabled like
```
@api.depends('channel_type')
def _compute_group_public_id(self):
    channels = self.filtered(lambda channel: channel.channel.type == 'channel')
    channels.group_public_id = False
    toupdate = (self - channels).filtered(lambda channel: not channel.group_public_id)
    toupdate.group_public_id = group_user
```

**Review by tde-banana-odoo — COMMENTED** [review]
Being pinged by bots, once again, bisous !

File: addons/mail/models/mail_channel.py (hunk 6 above)
Comment (seb-odoo) [reply addons/mail/models/mail_channel.py:108]: For the other check indeed it is not really necessary if we assume data are not messed up (and as we make the upgrade to clean them it should be fine), but there is no performance issue (we are talking about adding a simple condition on a field already in cache, in comparison to a method that already iterates over all the users of a group as well as all members of the channel).

But for data integrity having SQL constraint is always better than python when possible, as this is the actual source of truth. Python constraint assume the ORM is used, but it's not always the case. And it's probably more performant to let SQL handle it rather than having to run python code (I don't really know, to be confirmed by actual benchmark if necessary, but anyway I don't think the performance issue is a concern in either case here, as creating a new channel or updating these fields is very rarely done, and also never done on a lot of channels). So the SQL one I would do definitely it now, if there was only one thing to do in this PR, it would be that one.

File: addons/hr/models/mail_channel.py [hunk 11]
```diff
@@ -11,6 +12,14 @@ class Channel(models.Model):
         'hr.department', string='HR Departments',
         help='Automatically subscribe members of those departments to the channel.')
 
+    # CONSTRAINTS
+
+    @api.constrains('subscription_department_ids')
+    def _constraint_subscription_department_ids_channel(self):
+        failing_channel = self.sudo().filtered(lambda channel: channel.channel_type!='channel' and channel.subscription_department_ids)
```
Comment (seb-odoo) [inline addons/hr/models/mail_channel.py:19]: lint: spacing around operators

Comment (seb-odoo) [inline addons/hr/models/mail_channel.py:19]: name it `failing_channels` as there can be more than one

File: addons/hr/models/mail_channel.py [hunk 12]
```diff
@@ -11,6 +12,14 @@ class Channel(models.Model):
...
+        if failing_channel:
+            raise ValidationError(_("The following channels have department auto-subscription but are not of right type 'channel': %(channel_names)s"), channel_names = ', '.join(failing_channel.mapped('name')))
```
Comment (seb-odoo) [inline addons/hr/models/mail_channel.py:21]: `but are not of right type 'channel'` reads a bit strange. Either we should say they are not the right type in generic terms, or we should say what is the right type, but mixing both on the same expression looks incorrect. Maybe just using parenthesis instead of quote around channel would make it better.

Also note that channel name can actually be empty (in case of group with auto-name based on members), so this error might make no sense with just empty separated by `,`.

File: addons/mail/models/mail_channel.py [hunk 13]
```diff
@@ -88,21 +88,29 @@ def _generate_random_token(self):
...
 
     @api.constrains('channel_last_seen_partner_ids', 'channel_partner_ids')
     def _constraint_partners_chat(self):
         for ch in self.sudo().filtered(lambda ch: ch.channel_type == 'chat'):
             if len(ch.channel_last_seen_partner_ids) > 2 or len(ch.channel_partner_ids) > 2:
                 raise ValidationError(_("A channel of type 'chat' cannot have more than two users."))
 
+    @api.constrains('group_public_id', 'group_ids')
+    def _constraint_group_id_channel(self):
+        failing_channel = self.sudo().filtered(lambda channel: channel.channel_type!='channel' and (channel.group_public_id or channel.group_ids))
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:110]: lint: space around operators

Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:110]: `failing_channels`

File: addons/mail/models/mail_channel.py [hunk 14]
```diff
@@ -88,21 +88,29 @@ def _generate_random_token(self):
...
+        if failing_channel:
+            raise ValidationError(_("The following channels have group-based authorization but are not of right type 'channel': %(channel_names)s"), channel_names = ', '.join(failing_channel.mapped('name')))
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:112]: same remark

File: addons/mail/models/mail_channel.py [hunk 15]
```diff
@@ -168,6 +176,14 @@ def _compute_member_count(self):
         for channel in self:
             channel.member_count = member_count_by_channel_id.get(channel.id, 0)
 
+    @api.depends('channel_type')
+    def _compute_group_public_id(self):
+        group_user = self.env.ref('base.group_user')
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:181]: no need for variable if it's only used once now

File: addons/mail/models/mail_channel.py [hunk 16]
```diff
@@ -168,6 +176,14 @@ def _compute_member_count(self):
...
+        groups_chats = self.filtered(lambda channel: channel.channel_type != 'channel')
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:182]: `chats` is incorrect, as "not channel" contains more things than just chat.

I would invert the flow, to have a variable named `channels`, and do the difference the other way around.

File: addons/mail/views/mail_channel_views.xml [hunk 17]
```diff
@@ -73,7 +73,7 @@
                                 <group class="o_label_nowrap">
                                     <field name="public" widget="radio" string="Who can follow the group's activities?"/>
                                     <field name="group_public_id"
-                                        attrs="{'invisible': [('public','!=','groups')], 'required': [('public','=','groups')]}"
+                                        attrs="{'invisible': ['|', ('channel_type', '!=', 'channel'), ('public','!=','groups')], 'required': [('public','=','groups')]}"
                                         />
                                     <field name="group_ids" widget="many2many_tags"
```
Comment (seb-odoo) [inline addons/mail/views/mail_channel_views.xml:78]: invisible condition on channel type should also be added on `group_ids` (and also on department in hr)

File: addons/mail/models/mail_channel.py [hunk 18]
```diff
@@ -88,22 +88,30 @@ def _generate_random_token(self):
...
     def _constraint_partners_chat(self):
         for ch in self.sudo().filtered(lambda ch: ch.channel_type == 'chat'):
             if len(ch.channel_last_seen_partner_ids) > 2 or len(ch.channel_partner_ids) > 2:
                 raise ValidationError(_("A channel of type 'chat' cannot have more than two users."))
 
+    @api.constrains('group_public_id', 'group_ids')
+    def _constraint_group_id_channel(self):
+        failing_channels = self.sudo().filtered(lambda channel: channel.channel_type != 'channel' and (channel.group_public_id or channel.group_ids))
+        if failing_channels:
+            raise ValidationError(_("channel_type should be 'channel' to have the group-based authorization."))
```
Comment (seb-odoo) [inline addons/mail/models/mail_channel.py:113]: add failing channels in the error message to help debugging

File: addons/hr/models/mail_channel.py [hunk 19]
```diff
@@ -11,6 +12,12 @@ class Channel(models.Model):
         'hr.department', string='HR Departments',
         help='Automatically subscribe members of those departments to the channel.')
 
+    @api.constrains('subscription_department_ids')
+    def _constraint_subscription_department_ids_channel(self):
+        failing_channels = self.sudo().filtered(lambda channel: channel.channel_type != 'channel' and channel.subscription_department_ids)
+        if failing_channels:
+            raise ValidationError(_("channel_type should be 'channel' to have the department auto-subscription."))
```
Comment (seb-odoo) [inline addons/hr/models/mail_channel.py:19]: also failing_channels in error message

File: addons/mail/models/mail_channel.py (hunk 18 above)
Comment (seb-odoo) [reply addons/mail/models/mail_channel.py:113]: Also missing "or group auto-subscription" in the message (same idea as the SQL error)

File: addons/mail/tests/test_mail_channel.py [hunk 20]
```diff
@@ -242,6 +242,7 @@ def test_channel_recipients_chat(self):
         """ Posting a message on a chat should not send emails """
         self.test_channel.write({
             'channel_type': 'chat',
+            'public': 'private',
```
Comment (seb-odoo) [inline addons/mail/tests/test_mail_channel.py:245]: This diff is unrelated to the current PR?

**Comment by seb-odoo** [comment]
runbot issue

**Review by seb-odoo — APPROVED** [review]
@robodoo r+
